src/Common/src/Interop/Unix/System.Net.Security.Native/Interop.GssApiException.cs

```diff
@@ -0,0 +1,84 @@
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using Microsoft.Win32.SafeHandles;
+
+internal static partial class Interop
+{
+    internal static partial class NetSecurityNative
+    {
+        internal sealed class GssApiException : Exception
+        {
+            private Status _minorStatus;
+
+            public Status MinorStatus
+            {
+                get { return _minorStatus;  }
+            }
+
+            public GssApiException(string message) : base(message)
+            {
+            }
+
+            public GssApiException(Status majorStatus, Status minorStatus)
+                : this(SR.Format(SR.net_gssapi_operation_failed, majorStatus, minorStatus), majorStatus, minorStatus)
+            {
+            }
+
+            public GssApiException(string message, Status majorStatus, Status minorStatus)
+                : this(message)
+            {
+                HResult = (int)majorStatus;
+                _minorStatus = minorStatus;
+            }
+
+            public static void AssertOrThrowIfError(string message, Status majorStatus, Status minorStatus)
+            {
+                if (majorStatus != Status.GSS_S_COMPLETE)
+                {
+                    GssApiException ex = new GssApiException(majorStatus, minorStatus);
+                    throw ex;
+                }
+            }
+
+#if DEBUG
+            public override string ToString()
+            {
+                return Message + "\n GSSAPI status: " + GetGssApiDisplayStatus((Status)HResult, _minorStatus);
+            }
+
+            private static string GetGssApiDisplayStatus(Status majorStatus, Status minorStatus)
+            {
+                KeyValuePair<Status, bool>[] statusArr = { new KeyValuePair<Status, bool>(majorStatus, false), new KeyValuePair<Status, bool>(minorStatus, true) };
+                int length = statusArr.Length;
+                string[] msgStrings = new string[length];
+
+                for (int i = 0; i < length; i++)
+                {
+                    SafeGssBufferHandle msgBuffer;
+                    Interop.NetSecurityNative.Status minStat;
+                    int statusLength;
+                    if (Status.GSS_S_COMPLETE != DisplayStatus(out minStat, statusArr[i].Key, statusArr[i].Value, out msgBuffer, out statusLength))
+                    {
+                        msgStrings[i] = "Unknown Error";
+                        continue;
+                    }
+                    using (msgBuffer)
+                    {
+                        byte[] statusBytes = new byte[statusLength];
+                        msgBuffer.Copy(statusBytes, 0);
+                        var encoding = new System.Text.ASCIIEncoding();
+                        msgStrings[i] = encoding.GetString(statusBytes);
+                    }
+                }
+
+                return msgStrings[0] + " (" + msgStrings[1] + ") \nStatus: " + majorStatus.ToString("x8") + " (" + minorStatus.ToString("x8") + ")";
```

Comment:
I'd find this easier to read as a string.Format (and eventually as a $"..." string when we enable C# 6 in the repo).